oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/util/CreateMetricUpdater.java

```diff
@@ -0,0 +1,99 @@
+
+import org.apache.jackrabbit.oak.plugins.document.Collection;
+import org.apache.jackrabbit.oak.plugins.document.Document;
+import org.apache.jackrabbit.oak.plugins.document.DocumentStore;
+import org.apache.jackrabbit.oak.plugins.document.DocumentStoreStatsCollector;
+import org.apache.jackrabbit.oak.stats.MeterStats;
+import org.apache.jackrabbit.oak.stats.TimerStats;
+
+import java.util.List;
+import java.util.function.BiPredicate;
+import java.util.function.Predicate;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Abstract class to update the metrics for {@link DocumentStoreStatsCollector#doneCreate(long, Collection, List, boolean)} for underlying {@link DocumentStore}
+ *
+ * <p>Concrete implementations provide instances of {@link MeterStats}, {@link TimerStats} based on whether throttling is ongoing or not
+ */
+public abstract class CreateMetricUpdater {
+
+    private final MeterStats createNodeMeter;
+    private final MeterStats createSplitNodeMeter;
+    private final TimerStats createNodeTimer;
+    private final MeterStats createJournal;
+    private final TimerStats createJournalTimer;
+
+    public CreateMetricUpdater(final MeterStats createNodeMeter,
+                               final MeterStats createSplitNodeMeter,
+                               final TimerStats createNodeTimer,
+                               final MeterStats createJournal,
+                               final TimerStats createJournalTimer) {
+        this.createNodeMeter = createNodeMeter;
+        this.createSplitNodeMeter = createSplitNodeMeter;
+        this.createNodeTimer = createNodeTimer;
+        this.createJournal = createJournal;
+        this.createJournalTimer = createJournalTimer;
+    }
+
+    public void update(final Collection<? extends Document> collection, final long timeTakenNanos,
+                       final List<String> ids, final boolean insertSuccess,
+                       final BiPredicate<Collection<? extends Document>, Integer> isNodesCollectionUpdated,
+                       final CreateStatsConsumer<MeterStats, MeterStats, TimerStats> createStatsConsumer,
+                       final Predicate<Collection<? extends Document>> isJournalCollection,
+                       final StatsConsumer<MeterStats, TimerStats> journalStatsConsumer) {
+
+        requireNonNull(isNodesCollectionUpdated);
+        requireNonNull(isJournalCollection);
+        requireNonNull(createStatsConsumer);
+        requireNonNull(journalStatsConsumer);
+
+        if (isNodesCollectionUpdated.test(collection, ids.size()) && insertSuccess) {
+            createStatsConsumer.accept(createNodeMeter, createSplitNodeMeter, createNodeTimer, ids, timeTakenNanos);
+        } else if (isJournalCollection.test(collection)) {
+            journalStatsConsumer.accept(createJournal, createJournalTimer, ids.size(), timeTakenNanos);
+        }
+    }
+
+    public static class CreateMetricUpdaterWithoutThrottling extends CreateMetricUpdater {
```

Comment:
Why are there `CreateMetricUpdaterWithoutThrottling` and `CreateMetricUpdaterWithThrottling`? There seems to be no difference between the two, except for a different name. Similar for other updater related classes.

Reply:
I have created 2 implementations for each updater (abstract) class, one with & other without throttling.

Their implementations only differ in metrics that are passed into their respective constructors.
So the one with throttling has metrics with throttling and the other without (as is the case with their respective names). Rest everything is identical.

In this way, I can reuse the code in the base updater class, while using different metrics based on whether or not throttling is going on.

Reply:
> I can reuse the code in the base updater class, while using different metrics based on whether or not throttling is going on.

But you can do this with just `CreateMetricUpdater` if it wasn't abstract. I still don't see a need for the two sub classes.

Reply:
I get your point, will update the PR.